### PR TITLE
Implement client-side event retention

### PR DESCRIPTION
**Pull request by hawkowl**
Implementing the client-side requirements for https://github.com/wamp-proto/wamp-proto/commit/17c70b5acc8bab04f3573053a931def43c8679e6

### CODE REVIEW

**Comment by hawkowl** [comment]
@oberstet what were the cosmetics you wanted on this PR? I'm guessing the impl is all right, so I'll start work on the CB stuff assuming this is the API.

**Comment by oberstet** [comment]
@hawkowl "cosmetics": the parameter ordering I mentioned in the review comment above ..

**Comment by hawkowl** [comment]
@oberstet I don't see a review comment anywhere... I think you have to complete the review for it to show up?

**Comment by oberstet** [comment]
@hawkowl this is very confusing .. do you see it now?

**Comment by hawkowl** [comment]
@oberstet sure do, will fix

File: autobahn/wamp/message.py [hunk 1]
```diff
@@ -2023,7 +2065,8 @@ def parse(wmsg):
                     topic=topic,
                     enc_algo=enc_algo,
                     enc_key=enc_key,
-                    enc_serializer=enc_serializer)
+                    enc_serializer=enc_serializer,
+                    retained=retained)
```
Comment (hawkowl) [inline autobahn/wamp/message.py]: Okay, non-keyword based calling was my only concern. If we're not worried about that, I can relocate them to where it makes sense.

**Comment by oberstet** [comment]
@hawkowl great! rgd reviews only visible when review is submitted/finished: hopefully I remember that trap next time.

File: autobahn/wamp/message.py [hunk 2]
```diff
@@ -2067,7 +2113,7 @@ def __str__(self):
         """
         Returns string representation of this message.
         """
-        return u"Event(subscription={0}, publication={1}, args={2}, kwargs={3}, publisher={4}, publisher_authid={5}, publisher_authrole={6}, topic={7}, enc_algo={8}, enc_key={9}, enc_serializer={9}, payload={10})".format(self.subscription, self.publication, self.args, self.kwargs, self.publisher, self.publisher_authid, self.publisher_authrole, self.topic, self.enc_algo, self.enc_key, self.enc_serializer, b2a(self.payload))
+        return u"Event(subscription={0}, publication={1}, args={2}, kwargs={3}, publisher={4}, publisher_authid={5}, publisher_authrole={6}, topic={7}, enc_algo={8}, enc_key={9}, enc_serializer={9}, payload={10}, retained={11})".format(self.subscription, self.publication, self.args, self.kwargs, self.publisher, self.publisher_authid, self.publisher_authrole, self.topic, self.enc_algo, self.enc_key, self.enc_serializer, b2a(self.payload), self.retained)
```
Comment (hawkowl) [inline autobahn/wamp/message.py]: @oberstet cool, no problem. Do we want this on any of the other objects?

**Comment by hawkowl** [comment]
@oberstet yeah, I think it's so that it's one event & email, whilst the old way of doing it, the person whose PR it is would get like twenty emails trickling in... most of which don't make sense out of context.

File: autobahn/wamp/message.py (hunk 2 above)
Comment (oberstet) [inline autobahn/wamp/message.py]: I don't understand .. "get_retained" = SUBSCRIBE, "retained" => EVENT. what other WAMP messages do you mean?

**Comment by oberstet** [comment]
@hawkowl ah, ok. that is valid reason (not wanting to receive dozens of notifications) .. I see.

File: autobahn/wamp/message.py (hunk 2 above)
Comment (oberstet) [inline autobahn/wamp/message.py]: Ah, right. Yes, like you say: EVENT, SUBSCRIBE and PUBLISH have the flag(s), and in each case, the respective parameter order should be like I indicated for EVENT.

Comment (hawkowl) [inline autobahn/wamp/message.py]: Cool, can do.

**Comment by hawkowl** [comment]
@oberstet I think I got them all now? And in the ones where I had to re-arrange, I removed the Python 2.6 specific explicit `{}` numbering junk... don't have to support that anymore! :D

**Comment by oberstet** [comment]
perfect. yeah, the numbering stuff was annoying and error prone ..

**Pull request merged**